client/jmx-adapter/src/main/java/org/jolokia/client/jmxadapter/RemoteJmxAdapter.java

```diff
@@ -0,0 +1,586 @@
+    }
+
+    // special case, if the attribute is ObjectName
+    if (rawValue instanceof JSONObject
+        && ((JSONObject) rawValue).size() == 1
+        && ((JSONObject) rawValue).containsKey("objectName")) {
+      return getObjectName("" + ((JSONObject) rawValue).get("objectName"));
+    }
+    try {
+      return ToOpenTypeConverter.returnOpenTypedValue(qualifiedName, rawValue);
+    } catch (OpenDataException e) {
+      return rawValue;
+    }
+  }
+
+  private boolean isPrimitive(Object rawValue) {
+    return rawValue == null
+        || rawValue instanceof Number
+        || rawValue instanceof Boolean
+        || rawValue instanceof String
+        || rawValue instanceof Character;
+  }
+
+  static ObjectName getObjectName(String objectName) {
+    try {
+      return ObjectName.getInstance(objectName);
+    } catch (MalformedObjectNameException e) {
+      throw new UncheckedJmxAdapterException(e);
+    }
+  }
+
+  @Override
+  public AttributeList getAttributes(ObjectName name, String[] attributes)
+      throws InstanceNotFoundException, IOException {
+    AttributeList result = new AttributeList();
+
+    List<J4pReadRequest> requests = new ArrayList<J4pReadRequest>(attributes.length);
+
+    for (String attribute : attributes) {
+      requests.add(new J4pReadRequest(name, attribute));
+    }
+    List<?> responses = Collections.emptyList();
+    try {
+      responses = this.connector.execute(requests, this.defaultProcessingOptions());
+
+    } catch (J4pBulkRemoteException e) {
+      responses = e.getResults();
+    } catch (J4pException ignore) {
+      //will result in empty return
+    }
+    for (Object item : responses) {
+      if (item instanceof J4pReadResponse) {
+        J4pReadResponse value = (J4pReadResponse) item;
+        final String attribute = value.getRequest().getAttribute();
+        result.add(new Attribute(attribute,
+            adaptJsonToOptimalResponseValue(name, attribute, value.getValue())));
+      }
+    }
+    return result;
+  }
+
+  @Override
+  public void setAttribute(ObjectName name, Attribute attribute)
+      throws InstanceNotFoundException, AttributeNotFoundException, InvalidAttributeValueException,
+      IOException {
+    final J4pWriteRequest request =
+        new J4pWriteRequest(name, attribute.getName(), attribute.getValue());
+    try {
+      this.unwrapExecute(request);
+    } catch (UncheckedJmxAdapterException e) {
+      if (e.getCause() instanceof J4pRemoteException) {
+        J4pRemoteException remote = (J4pRemoteException) e.getCause();
+        if ("javax.management.AttributeNotFoundException".equals(remote.getErrorType())) {
+          throw new AttributeNotFoundException((e.getCause().getMessage()));
+        }
+        if ("java.lang.IllegalArgumentException".equals(remote.getErrorType())
+            && remote
+            .getMessage()
+            .matches(
+                "Error: java.lang.IllegalArgumentException : Invalid value .+ for attribute .+")) {
```

Comment:
Here I would also relax the regexp to match only on "Invalid value .+ attribute" to be more robust. Also, we can consider introducing an error enum to indicate like NOT_FOUND, INVALID_ATTRIBUTE which could be set accordingly on the backend (something for the future :)